src/Requests/MutationRequest.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Intermax\LaravelJsonApi\Requests;
+
+use Exception;
+use Illuminate\Contracts\Container\Container;
+use Illuminate\Foundation\Http\FormRequest;
+use Illuminate\Support\Collection;
+
+abstract class MutationRequest extends FormRequest
+{
+    abstract protected function type(): string;
+
+    public function authorize(): bool
+    {
+        if (! $this->hasHeader('Content-Type') || ! str_starts_with((string) $this->header('Content-Type'), 'application/vnd.api+json')) {
+            abort(415, 'Invalid content type');
+        }
+
+        return true;
+    }
+
+    /**
+     * @return array<string, mixed>
+     *
+     * @throws Exception
+     */
+    public function rules(Container $app): array
+    {
+        $rules = [
+            'data.id' => ['string'],
+            'data.type' => ['required', 'in:'.$this->type(), 'string'],
+        ];
+
+        // 'id' is required in JSON:API specification when updating a resource.
+        if ($this->method() !== 'POST') {
+            $rules['data.id'] = array_merge(['required'], $rules['data.id']);
+        }
+
+        $rules = $this->mergeIntoRules(
+            app: $app,
+            rules: $rules,
+            methodName: 'attributeRules',
+            pathPrefix: 'data.attributes'
+        );
+
+        return $this->mergeIntoRules(
+            app: $app,
+            rules: $rules,
+            methodName: 'relationshipRules',
+            pathPrefix: 'data.relationships'
+        );
+    }
+
+    /**
+     * @param  array<string, mixed>  $rules
+     * @return array<string, mixed>
+     *
+     * @throws Exception
+     */
+    public function mergeIntoRules(Container $app, array $rules, string $methodName, string $pathPrefix): array
+    {
+        $rulesCallable = [$this, $methodName];
+        // To allow the container to inject dependencies into the method that's being called
+        // we check if it exists through method_exists instead of defining a method to override.
```

Comment:
Would an abstract method be possible? Or an interface? That would help when implementing the custom methods

Reply:
No, unless we drop the dependency injection possibility. When overriding a method the signature should remain the same.

Reply:
Having the DI seems nice, but did we ever had to change the signature of the rules method?

Reply:
No, but I wanted to keep the features the same as the normal `rules` method (which also doesn't provide an abstract method for the same reasons).